### PR TITLE
Add monitoring of SageMaker APIs

### PR DESCRIPTION
**Pull request by ksaaskil**
- Expose new library APIs `meeshkan.monitor_sagemaker` and `meeshkan.sagemaker.monitor` (only leave one of these once decided)
- `meeshkan.sagemaker.monitor` returns an instance of `SageMakerJob`. I think this can be very useful if one can use the job instance to watch updates and do, e.g., `job.terminate()`.
- Added new `SageMakerJobMonitor` that creates `SageMakerJob`s and monitors them. This clearly overlaps with `tracker.py` and `TrackingPoller` therein, but I'd still like to keep them separate in the first phase and then refactor later on.
- Added `SageMakerHelper` that works as the adapter between `SageMakerJobMonitor` and `boto` client.  
- Create a new abstract base class `BaseJob`. Both `Job` and `SageMakerJob` inherit from this. I don't like inheritance that much but I guess it makes sense in this case as the base class is abstract and more like an "interface".
- `meeshkan.sagemaker.monitor` does not do anything yet, basically just prints status and sets everything up.
- TODO: `boto3` and `sagemaker` dependencies should probably be optional
- [x]  `boto` client is created at start time, which adds some overhead. Perhaps the client should only be created when `SageMakerJobMonitor` is first accessed and then cached? Is this the reason why starting tests fail? EDIT: Changed to lazy initialization.

### CODE REVIEW

**Comment by idantene** [comment]
I would naively suggest just extending the Executable API to a SageMakerExecutable and wrapping the various SM commands to fit our internal API. Do you think that could work?

**Comment by ksaaskil** [comment]
I don't think a SageMaker job is really an executable, is it not more like a `Trackable`, `Watchable` and/or `Stoppable`?

**Comment by ksaaskil** [comment]
Yeah, naming stuff needs a lot of care. I think right now `Executable` should not  be renamed to `Trackable` if it has methods like `launch_and_wait`, `terminate` or `stdout`. But this requires playing around with classes to see which abstractions would work.

**Comment by idantene** [comment]
@ksaaskil wouldn't `launch_and_wait` translate well to logically launching the monitor and waiting until it's done, etc?
Of course if you think it's better to have it as a sister of Job that's fine, I just feel that would add a lot of code duplication. Excited to see what you come up with though!

**Comment by ksaaskil** [comment]
In my understanding `Executable` would be anything executed by the Meeshkan agent. I think it's perhaps a bit large abstraction if `launch_and_wait` would mean both running Python scripts and monitoring a SageMaker job? I now created a base class `BaseJob` and made both `Job` (i.e. `SchedulerJob`) and `SageMakerJob` inherit from that. I'll work on this still next week and we can check what it would look like.

File: meeshkan/sagemaker/lib.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+        if sagemaker_job.status.is_processed:
+            print("Job {job_name} is already finished with status {status}.".format(job_name=sagemaker_job.name,
+                                                                                    status=sagemaker_job.status.name))
+        else:
+            print("Started monitoring job {job_name}, "
+                  "currently in phase {status}".format(job_name=sagemaker_job.name, status=sagemaker_job.status.name))
+        return sagemaker_job
+
+
+__all__ = ["monitor"]
```
Comment (idantene) [inline meeshkan/sagemaker/lib.py]: Better moved to the top of the file (after imports) for clarity and consistency with other modules

File: meeshkan/core/api.py [hunk 2]
```diff
@@ -140,6 +146,24 @@ def submit(self, args: Tuple[str, ...], cwd: str = None, name=None, poll_interva
...
+        :return: SageMakerJob instance
+        """
+        if not self.sagemaker_job_monitor:
+            raise RuntimeError("SageMaker job monitor not defined.")
+
+        job = self.sagemaker_job_monitor.create_job(job_name, poll_interval=poll_interval)
+        if job.status.is_processed:
+            return job
+        self.sagemaker_job_monitor.start(job)
+        # TODO self.job_store.include_as_job(job)
```
Comment (idantene) [inline meeshkan/core/api.py]: Is this `TODO` to maintain a persistent job store or to synchronize between Scheduler and SageMaker Monitor?

File: meeshkan/core/job.py [hunk 3]
```diff
@@ -50,11 +50,63 @@ def stale(self):
 SUCCESS_RETURN_CODE = [0]  # Completeness, extend later (i.e. consider > 0 return codes as success with message?)
 
 
-class Executable(object):
+class Trackable:
```
Comment (idantene) [inline meeshkan/core/job.py]: If we're adhering to Python3 only code (as we probably should), then maybe we should also remove other explicit inheritances from `object`; same for calls to `super()`). This is mostly my own doing, as I've used the Python2 syntax across the library.

File: meeshkan/core/job.py [hunk 4]
```diff
@@ -50,11 +50,63 @@ def stale(self):
...
+                 job_uuid: Optional[uuid.UUID] = None,
+                 job_number: Optional[int] = None,
+                 name: Optional[str] = None,
+                 poll_interval: Optional[float] = None):  # TODO Move also `status` here
+        super().__init__()
+        self.id = job_uuid or uuid.uuid4()  # pylint: disable=invalid-name
+        self.number = job_number  # Human-readable integer ID
+        self.poll_time = poll_interval or Job.DEF_POLLING_INTERVAL  # type: float
+        self.created = datetime.datetime.utcnow()
+        self.name = name or "Job #{number}".format(number=self.number)
```
Comment (idantene) [inline meeshkan/core/job.py]: Should a `BaseJob` also hold a status?

File: meeshkan/core/job.py [hunk 5]
```diff
@@ -155,11 +203,32 @@ def __repr__(self):
...
+class SageMakerJob(BaseJob):
+    """
+    Job run by SageMaker, meeshkan doing only monitoring.
+    """
+    def __init__(self,
+                 job_name: str,
+                 status: JobStatus,
+                 poll_interval: Optional[float]):
+        super().__init__(job_uuid=None,
+                         job_number=None,
```
Comment (idantene) [inline meeshkan/core/job.py]: If we use a `JobStore`, I guess `job_number` and/or `job_uuid` would be generated there. (Future TODO note, I guess)

File: meeshkan/core/job.py [hunk 6]
```diff
@@ -223,19 +287,9 @@ def __str__(self):
         return "Job: {executable}, #{number}, ({id}) - {status}".format(executable=self.executable, number=self.number,
                                                                         id=self.id, status=self.status.name)
 
-    def add_scalar_to_history(self, scalar_name, scalar_value) -> Optional[TrackerCondition]:
-        return self.scalar_history.add_tracked(scalar_name, scalar_value)
-
     def add_condition(self, *val_names, condition: Callable[[float], bool], only_relevant: bool):
```
Comment (idantene) [inline meeshkan/core/job.py]: Can't this also be moved to `BaseJob`?

File: meeshkan/core/sagemaker_monitor.py [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+
+
+LOGGER = logging.getLogger(__name__)
+
+
+# Do not expose anything by default (internal module)
+__all__ = []  # type: List[str]
+
+
+class SageMakerHelper:
```
Comment (idantene) [inline meeshkan/core/sagemaker_monitor.py]: Misses some typing (return and argument types)

File: meeshkan/core/sagemaker_monitor.py [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+            if self.connection_succeeded:
+                return
+            if not self.connection_succeeded:
+                raise SageMakerNotAvailableException
+
+        self.connection_tried = True
+
+        self.client = self.client or SageMakerHelper.build_client_or_none()
+
+        if not self.client:
```
Comment (idantene) [inline meeshkan/core/sagemaker_monitor.py]: Why not raise the Exception already in `build_client_or_none`?

File: tests/test_api.py [hunk 9]
```diff
@@ -230,3 +233,39 @@ def test_find_job_id_precedence(cleanup):  # pylint:disable=unused-argument,rede
...
+@pytest.fixture
+def mock_sagemaker_job_monitor():
+    return create_autospec(SageMakerJobMonitor).return_value
+
+
+@pytest.fixture
+def mock_api(mock_sagemaker_job_monitor):
+    scheduler = Scheduler(QueueProcessor())
+    service = create_autospec(Service).return_value
+    yield Api(scheduler=scheduler,
```
Comment (idantene) [inline tests/test_api.py]: Do we need a `yield` if there's no teardown after this point? (Or should `api.stop()` be called?)

File: tests/test_job.py [hunk 10]
```diff
@@ -48,4 +49,33 @@ def test_job_args_to_full_path_with_runtime():
...
+
+@pytest.fixture
+def example_job():
+    cwd, base = os.path.split(__file__)
+    return Job.create_job(args=(base, 'another argument'), cwd=cwd, job_number=1, name=JOB_NAME)
+
+
+def test_job_cancel_changes_status(example_job):
+    example_job.cancel()  # Runs with
+    assert example_job.status == JobStatus.CANCELLED_BY_USER
```
Comment (idantene) [inline tests/test_job.py]: Should we add assertion messages here and in the following tests? Or do you think it's trivial?

File: tests/test_job.py [hunk 11]
```diff
@@ -48,4 +49,33 @@ def test_job_args_to_full_path_with_runtime():
...
+    assert example_job.scalar_history is not None
+    assert example_job.name == JOB_NAME
+
+
+def test_job_add_tracked_works(example_job):
+    example_job.add_scalar_to_history("scalar", 0.1)
+
+
+def test_job_add_condition_works(example_job):
+    not example_job.add_condition("train loss", "evaluation loss",
```
Comment (idantene) [inline tests/test_job.py]: Without a `return` or `assert` statement, I'm guessing this (and `test_job_add_tracked_works`) are expected to run without raising any exceptions. That should probably be documented if that's the case :)
(Also the use of `not ...` without `assert` seems weird -- is it valid/important if the assumption is that nothing is raised?)

File: tests/test_sagemaker_monitor.py [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+        sagemaker_job_monitor.sagemaker_helper.get_job_status.return_value = JobStatus.QUEUED
+        job = sagemaker_job_monitor.create_job(job_name=job_name)
+        assert job.name == job_name
+        assert job.status == JobStatus.QUEUED
+
+    def test_create_running_job(self, sagemaker_job_monitor: SageMakerJobMonitor):
+        job_name = "foobar"
+        sagemaker_job_monitor.sagemaker_helper.get_job_status.return_value = JobStatus.RUNNING
+        job = sagemaker_job_monitor.create_job(job_name=job_name)
+        assert job.name == job_name
```
Comment (idantene) [inline tests/test_sagemaker_monitor.py]: Ditto regarding assertion messages

File: meeshkan/core/job.py (hunk 3 above)
Comment (ksaaskil) [reply meeshkan/core/job.py]: Yeah I thought here as we're supporting Python3 only we could use Python3-like class declarations.

File: meeshkan/core/sagemaker_monitor.py (hunk 8 above)
Comment (ksaaskil) [reply meeshkan/core/sagemaker_monitor.py]: I just thought this would be a bit more explicit.

File: tests/test_api.py (hunk 9 above)
Comment (ksaaskil) [reply tests/test_api.py]: I think `yield` is the standard syntax for defining what's returned from the pytest-fixture, it's for me a clear indication of what's run before a test, what's returned, and what's run after the test.

File: tests/test_job.py (hunk 10 above)
Comment (ksaaskil) [reply tests/test_job.py]: I think it's trivial for such a short test.

File: tests/test_job.py (hunk 11 above)
Comment (ksaaskil) [reply tests/test_job.py]: I can't remember why `not` is added there, probably it's leftover from some re-writing.

File: tests/test_api.py (hunk 9 above)
Comment (idantene) [reply tests/test_api.py]: Okay. I'm used to it being used solely as the latter and if there's no division for pre- and post-test, then `return` is used.

**Pull request merged**